### DISCUSSION TITLE
Question to snapshot.getPromise vs snapshot.getLoadable

### DISCUSSION THREAD

**Discussion by miracle0930** [facebookexperimental/Recoil#2012] [Q&A] [answered]
Hi team! Currently, I am confused with the ```snapshot.getPromise``` and ```snapshot.getLoadable``` method. To me these two methods are both used for getting a snapshot from an atom or a selector. The main difference is ```getPromise``` is async and ```getLoadable``` is sync. 

The way I am using them is quite the same, for example
```js
const useSomeCallback = () => {
    return useRecoilCallback(() => async ({ snapshot }) => {
        // use getPromise
        const aSnapshotFromPromise = await snapshot.getPromise(someAtomOrSelector)
        // use getLoadable
        const aSnapshotFromLoadable = snapshot.getLoadable(someAtomOrSelector).getValue()
        // some business logic...
    })
}
```
In most cases, both two work but sometimes it may different, sometimes ```getPromise``` may throw the ```got null or unexpected``` crash.

I am wondering what's the main difference between these two methods and when I should consider using ```getPromise``` and when I should consider using ```getLoadable```? 

Thanks in advance for your help!

**Comment by drarmstr** [comment, marked as answer]
As you point out they are very similar.  `getLoadable()` is synchronous and returns a `Loadable` object which encodes the state and either the value, error, or promise to an async value.  `getPromise()` on the other hand returns an async promise to the value, even if the value is synchronous or rejected to the error.  You may wish to use the `getLoadable()` version if you want to synchronously handle the result, error, or pending state.
